ARC-60 Algorand Wallet Structured Data Signing API

deanstef opened the pull request:
An API for signing structured bytes with Algorand wallets. It unlocks many use cases ranging from self-custodial digital identities, authentication, document signing etc.

The API provides a standard interface for a signData() function that takes as an argument a structured data object and metadata and returns the requested digital signature. It is compatible with ~~rekeyed accounts,~~ multi-signature and HD wallets.

This API is inspired and based on [ARC-1](https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0001.md). It has been conceived to be secure by design with semantics and security controls. It does not allow the signing of canonical Algorand transactions or Groups.

ehanoc commented on assets/arc-0060/types.ts:
```diff
@@ -0,0 +1,33 @@
+    LSIG,
+  }
+
+export type StdData = string
```
I would avoid strings as they can have a lot of different representations. I would use `Uint8Array`

deanstef replied:
`strings` types are supported by [JSON Schema](https://json-schema.org/understanding-json-schema/reference/type). Having bytes would make ARC-60 interoperability less straightforward.

Note that this approach follows the same rationale as in [ARC-1](https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0001.md#rationale)

ehanoc replied:
But strings can be encoded, dynamic sizing and there come be ways of going around the schema verification if you double encode the data to be signed. for example. 

I would highly discourage using string to represent "raw" bytes. 

Instead, we can use JSONSchema to enforce byte level value range and how many exactly we want to sign. For example, a 32 byte challenge for the purpose of auth; could be represented with: 

```json
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$id": "http://arc60/schemas/auth.request.json",
    "type": "object",
    "additionalProperties": false,
    "properties": {
        "0": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "1": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "2": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "3": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "4": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "5": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "6": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "7": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "8": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "9": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "10": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "11": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "12": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "13": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "14": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "15": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "16": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "17": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "18": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "19": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "20": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "21": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "22": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "23": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "24": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "25": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "26": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "27": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "28": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "29": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "30": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        },
        "31": {
            "type": "integer",
            "minimum": 0,
            "maximum": 255
        }
    },
    "required": [
        "0",
        "1",
        "2",
        "3",
        "4",
        "5",
        "6",
        "7",
        "8",
        "9",
        "10",
        "11",
        "12",
        "13",
        "14",
        "15",
        "16",
        "17",
        "18",
        "19",
        "20",
        "21",
        "22",
        "23",
        "24",
        "25",
        "26",
        "27",
        "28",
        "29",
        "30",
        "31"
    ]
}
```

deanstef replied:
Thank you @ehanoc . 

The schema can be used for passing extra information with the signing bytes. For example, in the authentication process you might want to add the verifier domain or additional data. This is the reason why I added to the ARC-60 schema the `additionalProperties = true`.

Regarding your point, the schema has a required property `bytes` that represent the bytes to be signed. This property can be a byte array as you suggest. I updated the schema to reflect that: https://github.com/deanstef/ARCs/blob/7447a0d82b64677267fdd7e404d5bfc0ff54c5cf/assets/arc-0060/simple-schema.json#L8

deanstef replied:
I am resolving this comment now since we implemented an example for `AUTH`. Feel free to re-open it for further discussion.

ehanoc commented on assets/arc-0060/types.ts:
```diff
@@ -0,0 +1,33 @@
+  }
+
+export type StdData = string
+export type Ed25519Pk = string
```
Same here as `Uint8Array`, specially being Key material

deanstef replied:
Addressed here https://github.com/algorandfoundation/ARCs/pull/284/commits/7447a0d82b64677267fdd7e404d5bfc0ff54c5cf

ehanoc commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,506 @@
+
+#### Interface `Ed25519Pk`
+
+An `Ed25519Pk` object is a 32-byte public key, point of the ed25519 elliptic curve. It is base64 encoded in accordance with <a href="https://www.ietf.org/rfc/rfc7468.txt">RFC-7468</a>. The key is not transformed into an Algorand address.
```
I would use keys in their raw form. 32 bytes, skip the encoding, specially when this adds size.

ehanoc commented on ARCs/arc-0060.md:
````diff
@@ -0,0 +1,506 @@
+
+```tsx
+export type SignDataFunction = {
+    arbData: StdSignData[],
````
Why does it need to bulk sign? Most Key Management Systems do not support bulk signing and i can't see the purpose of doing it here. 

If the client needs multiple signatures, if it should be done multiple times. So it's also authorized, checked, logged multiple times

deanstef replied:
Good point. Although I think it is more friendly for users to have bulk signs (e.g. signing documents), I also agree that we should mimic the state-of-the-art signing approaches like KMS/HSMs. I will change it to single signing.

deanstef replied:
https://github.com/algorandfoundation/ARCs/pull/284/commits/67754a89ff9e211cfee6f29ebd7aa4f016b19243

ehanoc commented on ARCs/arc-0060.md:
````diff
@@ -0,0 +1,506 @@
+`SignedDataStr` is the produced 64-byte array ed25519 digital signature of the `StdData` object.
+
+```tsx
+export type SignedDataStr = string;
````
64 byte array instead of a string? `Uint8Array`

deanstef replied:
Addressed here https://github.com/algorandfoundation/ARCs/pull/284/commits/7447a0d82b64677267fdd7e404d5bfc0ff54c5cf

ehanoc commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,506 @@
+
+| ScopeType | Description |
+| --- | --- |
+| ARBITRARY | Signature of an array of arbitrary bytes. This is the most generic scope. |
```
Would it make sense to add the Scopes now? i.e `Authorization`, `LSig`, etc ? 

This way, depending on the Scope the wallet knows which schema to validate against

deanstef replied:
I think it makes sens for `Lsig`. For Auth, I would introduce the new `ScopeType` with ARC-31.

deanstef replied:
Resolving since we have `AUTH` and `LSIG` scopes (+ schemas) now.

ehanoc commented on assets/arc-0060/wallet.ts:
We could add a matching wallet.spec.ts file with a small test suite so that we can test the functions of the reference implementation

ehanoc commented on assets/arc-0060/lsig-schema.json:
@joe-p can you validate this schema?

ehanoc reviewed (commented):
Made a few comments. 

I think we need a schema for 32 bytes signing auth challenge requests

deanstef commented:
> Made a few comments.
> 
> I think we need a schema for 32 bytes signing auth challenge requests

Thank you for your comments. The JSON schema for authentication challenge will be provided by ARC-31 as an extension of ARC-60.

deanstef commented:
With https://github.com/algorandfoundation/ARCs/pull/284/commits/7447a0d82b64677267fdd7e404d5bfc0ff54c5cf I updated ARC-60 with a simpler signBytes` interface. It now takes three arguments: `data`, `metadata`, and `signer`. I removed the `msig`, `authaddr` and `hdWallet` functionalities to keep it simple.

@ehanoc @pbennett @yigitguler @k13n @joe-p PTAL

ehanoc commented on assets/arc-0060/wallet.ts:
```diff
@@ -0,0 +1,79 @@
+}
+
+const signedBytes = await signData(arc60Data, metadataMock, signerPk)
+if (!(signedBytes === null)) {
```
Can we verify the signature instead?

deanstef replied:
I have added the signature verification now :)

ehanoc commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+
+| ScopeType | Description |
+| --- | --- |
+| PLAIN | Signature of a simple byte string. This is the most generic scope. |
```
PLAIN is the same as ARBITRARY, doesn't specify **WHY** we are signing

deanstef replied:
Agreed. I would like to keep a generic Scope for a generic message signing. What about `SIGMSG`?

ehanoc replied:
But generic is problematic because we don't know what we are signing. `SIGMSG` still doesn't state what we are signing.

deanstef replied:
I removed generic scopes. We only have `AUTH` and `LSIG` now.

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+- `data` **MUST** be validated with respect to the JSON schema `metadata.schema`. If the validation fails, the call **MUST** be rejected with a `4300` error.
+- if the encoding `metadata.encoding` is present, it **MUST** be used to encode `data`.
+- the wallet **MUST** prepare the signing data as `signData=(<data.bytes>)`.
+  - if the value of `ARC6Domain` is not empty than it must used as `prefix`: `signData=(<prefix> || <data.bytes>)`.
```
typo: `ARC60Domain`

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+  - if the value of `ARC6Domain` is not empty than it must used as `prefix`: `signData=(<prefix> || <data.bytes>)`.
+- the wallet **MUST** ask users for signing confirmation. It **MUST** display the `metadata.message` if present, and the `signData` being signed:
+  - if the user approves, `signData` **MUST** be signed with `signer` and `ret` **MUST** be set to the corresponding `SignedDataStr`.
+  - if the user rejects, the call **MUST** fail with error code `4001`.
```
above it was indicated that rejecting means:
> **Rejecting** means throwing an error with `err.code=4300`

Here it says we should error code 4001

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+    /**
+    * Optional encoding used to represent the signing data.
+    */
+    encoding?: string;
```
What are possible encodings that should be supported?

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+
+#### Interface `SignedDataStr`
+
+`SignedDataStr` is the produced 64-byte array, ed25519 digital signature, of the signed data.
```
According to this definition, `SigendDataStr` is purely the signature (64 bytes long). In the reference implementation, [`nacl.sign`](https://www.npmjs.com/package/tweetnacl#naclsignmessage-secretkey) is used which returns the signature _and_ the original message.

If `SignedDataStr` represents purely the signature, wouldn't something like `Signature` be a more appropriate name for this type?

deanstef replied:
Good catch. Here https://github.com/algorandfoundation/ARCs/pull/284/commits/b244ed632ed393a0b33fea1fb0f6c4de03e74506 I updated both the naming and the reference implementation. I am using [`nacl.sign.detached`](https://www.npmjs.com/package/tweetnacl#naclsigndetachedmessage-secretkey) now.

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+    /**
+    * Canonical representation of the JSON schema for the signing data.
+    */
+    schema: string;
```
Why is the schema (which is a JSON object) represented as a canonicalized JSON string?

deanstef replied:
In a very first version of ARC60 I was using canonicalized versions of JSON objects to produce signatures of the entire objects. In the new version this is not needed anymore. We can get rid of the canonicalized stuff indeed.

deanstef replied:
https://github.com/algorandfoundation/ARCs/pull/284/commits/045de9ecaeb4247f881761e365bc8c868380df65

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+    /**
+    * Optional encoding used to represent the signing data.
+    */
+    encoding?: string;
```
What are possible encodings that should be supported?

jasonpaulos replied:
+1 to clarifying this

deanstef replied:
It can be any encoding. If the wallet does not support it, it will just reject the call. I make it explicit here https://github.com/algorandfoundation/ARCs/pull/284/commits/88923d3723674870d1f442e643f02834a99afb3e

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+
+##### Simple Message Data JSON Schema
+
+The JSON schema for simple message signing. The `StdData` **MUST** be an object of this schema.
```
Could `ScopeType.MSGSIG` be referenced here to make it explicit that this section is intended for that particular scope?

deanstef replied:
I am resolving this comment since we only have an example for `ScopeType.AUTH` now https://www.npmjs.com/package/tweetnacl#naclsigndetachedmessage-secretkey

k13n commented on ARCs/arc-0060.md:
````diff
@@ -0,0 +1,330 @@
+
+> The signing data JSON schema is inspired by the schema proposed with <a href="https://eips.ethereum.org/EIPS/eip-712">EIP-712: Typed structured data hashing and signing proposal</a>.
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "ARC60Domain": {
+      "type": "string",
+      "description": "The ARC-60 domain separator"
+    },
+    "bytes": {
+      "type": "string",
+      "description": "Byte string being signed. Used with ScopeType==MSGSIG."
+    },
+  "required": ["ARC60Domain", "bytes"]
+}
````
Is this outdated here? It seems in the file `simple-schema.json` you replaced the `bytes` (here type string) with an integer array as per @ehanoc's suggestion.

k13n commented on ARCs/arc-0060.md:
````diff
@@ -0,0 +1,330 @@
+
+For example, a valid object that takes as bytes an extra property field is:
+
+```json
+{
+  "ARC60Domain" : "arc60",
+  "bytes" : "ARC-60 is awesome"
+}
+```
````
Also here `bytes` might be an integer array?

deanstef replied:
Same as above -> https://github.com/algorandfoundation/ARCs/pull/284/commits/b244ed632ed393a0b33fea1fb0f6c4de03e74506

k13n commented on ARCs/arc-0060.md:
````diff
@@ -0,0 +1,330 @@
+The `SignDataError` object extends the `SignTxnsError` defined in [ARC-1](./arc-0001.md).
+
+```ts
+export interface SignDataError extends Error {
+  code: number;
+  data?: any;
+  failingSignData: (StdData | null);
+}
````
In which case could `failingSignData` be `null`?

deanstef replied:
For exampole, if the `data` object passed to the `signData(data, metadata, signer)` function was `null`.

deanstef replied:
If the error was caused by a missing StdData object.

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+
+### Semantic Requirements
+
+The call `signData(data, metadata, signer)` **MUST** either return the signed data `ret` or reject the call throwing an error `err`.
```
According to the definition of `SignDataFunction` above, it seems a `null` return value is also possible (which I don't think is required as I mentioned earlier).

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+- `data` **MUST** be validated with respect to the JSON schema `metadata.schema`. If the validation fails, the call **MUST** be rejected with a `4300` error.
+- if the encoding `metadata.encoding` is present, it **MUST** be used to encode `data`.
+- the wallet **MUST** prepare the signing data as `signData=(<data.bytes>)`.
+  - if the value of `ARC6Domain` is not empty than it must used as `prefix`: `signData=(<prefix> || <data.bytes>)`.
```
typo: `ARC60Domain`

deanstef replied:
https://github.com/algorandfoundation/ARCs/pull/284/commits/a8de5d08f3c558bedeb8c1addbacc25362eda7b5

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+Upon calling `signData(data, metadata, signer)`:
+
+- the `data`, `metadata`, and `signer` **MUST NOT** be `null`, otherwise the wallet **MUST** reject the call.
+- `data` **MUST** be validated with respect to the JSON schema `metadata.schema`. If the validation fails, the call **MUST** be rejected with a `4300` error.
```
According to the table above an invalid schema is a 4601 error code, or is this a different schema?

deanstef replied:
Error code 4601 means that the JSON schema does not comply with ARC-60 (the required fields `ARC60Domain` and `bytes` are missing).

In this case, we are failing because `data` is not a valid object according to the schema `metadata.schema`.

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+  - the wallet **MUST** reject if `ARC60Domain="TX"`.
+  - the wallet **MUST** reject if `ARC60Domain="TG"`.
+  - the wallet **MUST** reject if `metadata.scope=MSGSIG` and `ARC60Domain` field is not set to `"arc60"` or `""`.
+- the `bytes` must be a valid byte string:
+  - the wallet **MUST** reject if `bytes` is prepended with a forbidden domain separator `TX` or `TG`.
```
Is this validation here independent of the `ARC60Domain` validation? Because if `ARC60Domain` is set to `arc60` (or anything other than `""`, `"TX"`, or `"TG"`), wouldn't it be fine if `bytes` started with `TX` or `TG`? Because the wallet would have to concatenate the value of `ARC60Domain` and `bytes` and hence no malicious payload could be signed?

deanstef replied:
Setting `ARC60Domain` to `""` would make the concatenation not effective. I suggest to always validate `bytes` against malicious payloads (prepended with `TX` or `TG`) to avoid problems.

k13n reviewed (commented):
Thanks @deanstef for putting together ARC-60!

ehanoc commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+
+| ScopeType | Description |
+| --- | --- |
+| MSGSIG | Signature of a simple message. This is the most generic scope. |
```
I don't think we should have generic scopes. They should be VERY Specific. 

A generic scope doesn't tell you what schema to enforce and could potentially be used to trick the user or even the wallet.

FrankSzendzielarz replied:
The spec doesn't make it generic. It says that the bytes must not be certain things, such as transactions. The spec says "most generic", i.e., that narrower scopes might be defined. I think the problem then becomes how to specify what things should *not* be. For example, specifying a scope that allows 32 arbitrary bytes to be signed for auth purposes does not preclude a 32 byte arbitrary document from being signed.

deanstef replied:
We only have `AUTH` and `LSIG` scopes now.

ehanoc commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+
+> This section is non-normative
+
+Signatures of bytes are processed with the `signData(data, metadata, signer)` function.
```
@deanstef requiring the public key (a.k.a signer) would be somewhat cumbersome for HDWallets to find the correct address to use. Would have to do a recursive search. 

Legacy algorand wallets are single address, with HDWallets they will support BIP44 and providing the path  as optional would suit best.

Thoughts?

deanstef replied:
In the first version of ARC-60 there was the possibility of passing a HDWallet derivation path as signer parameter (see https://github.com/deanstef/ARCs/blob/67754a89ff9e211cfee6f29ebd7aa4f016b19243/ARCs/arc-0060.md?plain=1#L122)

Then we decided to keep it simple and just opt for passing the public key of the signer. I'd be happy to re-introduce HDWallets support though

ehanoc replied:
I understand. Although just the public key is ambiguous when you have N possible key pairs in a BIP32 / BIP44 path scheme.

deanstef replied:
This has been addressed

deanstef commented on ARCs/arc-0060.md:
````diff
@@ -0,0 +1,330 @@
+`SignedDataStr` is the produced 64-byte array, ed25519 digital signature, of the signed data.
+
+```tsx
+export type SignedDataStr = Uint8Array;
````
https://github.com/algorandfoundation/ARCs/pull/284/commits/045de9ecaeb4247f881761e365bc8c868380df65

deanstef commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,330 @@
+    metadata: StdSignMetadata,
+    signer: Ed25519Pk,
+}
+    => Promise<(SignedDataStr | null)>;
```
Yes you are right. This is a leftover from a previous design with bulk signatures.

deanstef replied:
Actually - in the reference implemnetation, I am returning `null` if the user does not confimr the signature: https://github.com/deanstef/ARCs/blob/a8de5d08f3c558bedeb8c1addbacc25362eda7b5/assets/arc-0060/wallet.ts#L79C4-L79C5

jasonpaulos commented on ARCs/arc-0060.md:
````diff
@@ -0,0 +1,328 @@
+The `StdDataStr` object is a string that comply with the `metadata` JSON schema.
+
+```tsx
+export type StdDataStr = string;
````
I don't understand how this can contain an logicsig program if it's a string

mxmauro replied:
Yes, I prefer a Buffer or Uint8Array.

FrankSzendzielarz replied:
The way I read the spec is that the StdDataStr is a stringified JSON object that must comply with the metadata supplied as an argument to sign function, of type StdSignMetadata. The metadata specifies which field contains data to be signed, and not the whole json object. Part of the problem is this line in the first paragraphs:  

> data is a string representing the structured data being signed. 

This is not true according to the rest of the spec. The data being signed is whichever field of the StdDataStr json to be signed. 

Another problem is that there is not defined here the standard encoding for JSON bytes data. Typically in web applications it is Base64 encoded strings, and OpenAPI spec clients/servers will work with that, but it's only by common convention, and I don't know the base specs like JSON Schema say anything about it (off the top of my head)

In general I find the spec confusing because of a) the use of both TypeScript types and JSON types, and I don't like the use of things like UInt8Array at all, as they are TS domain specific concepts and don't belong in language agnostic specs. b) slightly misleading terminology, like "Scope" instead of "Schema"...

deanstef replied:
Thank you @FrankSzendzielarz 

This commit https://github.com/algorandfoundation/ARCs/pull/284/commits/c08b9958f339073ce1711f080e7e34a7af3b940e should fix the issue you raised.

deanstef replied:
Also, @jasonpaulos @mxmauro please refer to the [Backwards Compatibility](https://github.com/deanstef/ARCs/blob/byte-signing/ARCs/arc-0060.md#backwards-compatibility) section of the ARC.

FrankSzendzielarz commented:
> Hi, I cannot figure out if, on the latest revision, data to sign is `Uint8Array` or a `string`. (I prefer the former because data & platform-independant)

The data to sign is an encoded field of the input JSON text. It will always be a string, of course, as the input is JSON. What encoding depends on the metadata (and I am informed that the spec will be updated to clarify this). What does seem to be clear though, given the comments like this one, is that either the spec is not being read, or not being understood, so I'd suggest specs like these get plain English TLDRs at the top.

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,449 @@
+- the `signingData` and `metadata` **MUST NOT** be `null`, otherwise the wallet **MUST** reject the call.
+- `signingData.data` **MUST** be valid respect the JSON schema in `metadata.schema`. If the validation fails, the call **MUST** be rejected with a `4300` error.
+- if the encoding `metadata.encoding` is present, it **MUST** be used to decode `signingData.data.bytes`.
+- the wallet **MUST** prepare the signing message as `msg=(<signingData.data.bytes>)`.
+  - if the value of `ARC60Domain` is not empty than it **MUST** be used as `prefix`: `msg=(<prefix> || <signingData.data.bytes>)`.
```
Can we generalize / simplify this statement and just write:

"the wallet **MUST** prepare the signing message as `msg=(<signingData.data.ARC60Domain> || <signingData.data.bytes>)`."

If `ARC60Domain` is empty, it will anyway be equivalent to `msg=(<signingData.data.bytes>)`

deanstef replied:
https://github.com/algorandfoundation/ARCs/pull/284/commits/4bad240350f70b931991c4765ea7ea0886999b52

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,449 @@
+- if the encoding `metadata.encoding` is present, it **MUST** be used to decode `signingData.data.bytes`.
+- the wallet **MUST** prepare the signing message as `msg=(<signingData.data.bytes>)`.
+  - if the value of `ARC60Domain` is not empty than it **MUST** be used as `prefix`: `msg=(<prefix> || <signingData.data.bytes>)`.
+- the wallet **MUST** ask users for signing confirmation. It **MUST** display a message (defined in Display message) providing the `ScopeType`, `signer`, and the `msg` being signed:
```
The Display message doesn't exist anymore. Can we rephrase and write something like:

"the wallet **MUST** ask users for signing confirmation. It **MUST** display a message providing the `ScopeType`, `signer`, and the `msg` being signed:"

deanstef replied:
There is a section "Display Message" that denscribes what you said (https://github.com/algorandfoundation/ARCs/pull/284/commits/83a4b793397158c52b83a5b3a1ac058e0198e2e4)

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,449 @@
+  - it **MUST** be a valid `StdDataStr` object (UTF-8 encoded), otherwise the wallet **MUST** reject.
+  - it **MUST** adhere to `metadata.schema`, otherwise the wallet **MUST** reject.
+  - the `data.ARC60Domain` field must be validated against forbidden values:
+    - the wallet **MUST** reject if `ARC60Domain="TX"`.
+    - the wallet **MUST** reject if `ARC60Domain="TG"`.
```
Same as above: are there any other forbidden tags like "Program"?

deanstef replied:
https://github.com/algorandfoundation/ARCs/pull/284/commits/19720ff8a4b186fedd7f9884a3d97400d6f94780

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,449 @@
+  - `data.bytes` must be a valid byte string:
+    - it **MUST** be encoded as `metadata.encoding` if present, otherwise `base64`.
+    - the wallet **MUST** throw a `4602` error if the decoding fails with the given encoding value.
+    - the wallet **MUST** reject if `data.bytes` field is prepended with a forbidden domain separator `TX` or `TG`.
```
Just marking this for completeness -- same as above

deanstef replied:
https://github.com/algorandfoundation/ARCs/pull/284/commits/19720ff8a4b186fedd7f9884a3d97400d6f94780

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,449 @@
+
+This API was not designed to sign Algorand transactions or a group of transactions (see ARC-1 instead).
+
+## Backwards Compatibility
```
I'm wondering if we should even mention ARC-47 in this ARC here. I'm inclined to remove it for the sake of simplicity and to keep this ARC as short and concise as possible. If there's demand for signing lsigs with ARC-60, somebody can propose a new ARC, specify the scope, and define the schema, etc.

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,449 @@
+
+Users must be aware of what they are signing and for what purpose. The wallet **MUST** always compute a new message dynamically according to the received `signingData` and `metadata`. This prevents the injection of malicious or misleading messages and provides a clear scope of the signing action.
+
+Wallets only support known `ScopeTypes` and schemas. It ensures that signing actions have a clear scope and mitigates the risk of producing unexpected signatures.
```
I agree that wallets need to hard-code the `ScopeTypes` and schemas that they support. But this raises the question for me if we even need to include the `schema` in the `metadata` field?

Given the `ScopeType`, the wallet could look up its hard-coded schema and match the `signingData` against this schema.

Perhaps it's nice to have the schema also in the metadata to make the request self-descriptive. But in an implementation I'd likely ignore `metadata.schema` and match `signingData.bytes` against my hard-coded schema. Would that be a problem?

k13n commented on ARCs/arc-0060.md:
````diff
@@ -0,0 +1,449 @@
+}
+```
+
+- The `ARC60Domain` object indicates the domain separator to be used for signing. It **SHOULD** be set to `"arc60"` or empty `""`. Algorand transaction domain separators `TX` and `TG` **MUST** be forbidden.
+- The `bytes` object is the byte array of data being signed. In the example, it is a 32-byte challenge to be signed for authentication scope. The `bytes` field cannot be a valid Algorand transaction, therefore it **MUST NOT** be prepended with a known domain separator `TX` or `TG`.
````
Is ruling out certain prefixes in the signing `bytes` a good idea? 

Assume an application creates a random 32-byte string that needs to be signed as a challenge. If that byte string starts with `TX`, `TG`, or any other forbidden prefix, the wallet must reject it.  This may be confusing to an (unaware) application: most of the time the wallet is accepting its challenge but sometimes it is refusing to sign the data because the random data happens to have a forbidden prefix.

Alternatively, what if we force `ARC60Domain` to be non-empty and to not be one of a set of blacklisted prefixes (`TX`, `TG`, ...)? Then we wouldn't have to restrict the `bytes`.

deanstef replied:
1. `Program` should be forbidden only **IF** `ScopeType` != `LSIG`. I am clarifying this in the semantics as well. And yes, `TG` is for groups https://github.com/algorand/go-algorand/blob/981abad60d1fdd2993847434417982b4d373b7b8/protocol/hash.go#L74 
2. This problem is difficult to resolve and maybe it is a UX/UI issue. Auth protocols working with a `challenge-response` have no idea of our prefix-thing. Your proposal would break the compatibility with those protocols and I don't like it. For example see [here](https://www.w3.org/TR/webauthn-2/#sctn-sample-authentication)

k13n commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,449 @@
+  - the wallet **MUST** follow the ARC-47 specification: compile the program from the received template and check its integrity with the `hash`.
+  - the compiled program **MUST** be used with the prefix to compute the signing message `msg`, as `msg="Program" || <compiled_lsig>`.
+
+## Test Cases
```
Once the ARC is stable and about to be finalized, it'd be nice to add some test cases. We could specify a keypair, a message (according to the AUTH schema) and show the resulting signature as a test case.

In another test case we could have a keypair derived from a hd-wallet.

deanstef commented on ARCs/arc-0060.md:
````diff
@@ -0,0 +1,449 @@
+```json
+{
+  "ARC60Domain" : "arc60",
+  "bytes" : "[99, 50, 97, 57, 51, 99, ..., 57, 49, 53]"
````
https://github.com/algorandfoundation/ARCs/pull/284/commits/028580510f3b78fe9942483a0246c9632e21166d

deanstef commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,449 @@
+
+`signingData` is a `StdSigData` object composed of the signing `data` that instantiates a known JSON Schema and the `signer`'s public key.
+
+`metadata` is a `StdSignMetadata` object that describes the signature `scope`, the JSON `schema`, and the `encoding` used to represent the signing bytes. The JSON schema takes two elements: the `bytes` being signed and an optional `prefix`.
```
https://github.com/algorandfoundation/ARCs/pull/284/commits/f992d9ec22ccc49363eb118adbc8964052571c9c

deanstef commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,449 @@
+  "scope": "...",
+  "schema": "/{.../}",
+  "encoding": "..."
+}
```
IRL we agree that both`data` and `schema` can be JSON objects rather than strings

https://github.com/algorandfoundation/ARCs/pull/284/commits/43a7178ae4a37d71eb78abf7cd4049f40ea0d12a

deanstef commented on ARCs/arc-0060.md:
````diff
@@ -0,0 +1,449 @@
+  "schema": "/{.../}",
+  "encoding": "..."
+}
+```
````
https://github.com/algorandfoundation/ARCs/pull/284/commits/43a7178ae4a37d71eb78abf7cd4049f40ea0d12a

deanstef commented on ARCs/arc-0060.md:
```diff
@@ -0,0 +1,449 @@
+
+> Following [ARC-1](./arc-0001.md) terminology, in this ARC the term **Rejecting** means throwing an error with `err.code=4300`.
+
+Upon calling `signData(signingData,, metadata)`:
```
https://github.com/algorandfoundation/ARCs/pull/284/commits/5c4184e6ee6928f7ad106ba1daae44e3e02b1f9b

SudoWeezy commented:
Closing for https://github.com/algorandfoundation/ARCs/pull/313
